hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMBucketLayoutUpgrade.java

```diff
@@ -0,0 +1,272 @@
+import org.apache.hadoop.ozone.om.helpers.OmVolumeArgs;
+import org.apache.hadoop.ozone.om.protocol.OzoneManagerProtocol;
+import org.apache.hadoop.ozone.om.upgrade.OMLayoutFeature;
+import org.apache.hadoop.ozone.upgrade.UpgradeFinalizer;
+import org.apache.ozone.test.LambdaTestUtils;
+import org.junit.Test;
+import org.junit.Before;
+import org.junit.After;
+import org.junit.Rule;
+import org.junit.rules.Timeout;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.UUID;
+
+import static org.apache.hadoop.ozone.OzoneConsts.LAYOUT_VERSION_KEY;
+import static org.apache.hadoop.ozone.om.OMUpgradeTestUtils.waitForFinalization;
+import static org.apache.hadoop.ozone.om.upgrade.OMLayoutFeature.INITIAL_VERSION;
+import static org.apache.hadoop.ozone.om.upgrade.OMLayoutVersionManager.maxLayoutVersion;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.fail;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Upgrade testing for Bucket Layout Feature.
+ * <p>
+ * Expected behavior:
+ * 1. Pre-Finalize: OM should not allow creation of buckets with new bucket
+ * layouts. Only LEGACY buckets are allowed.
+ * <p>
+ * 2. Post-Finalize: OM should allow creation of buckets with new bucket
+ * layouts.
+ */
+@RunWith(Parameterized.class)
+public class TestOMBucketLayoutUpgrade {
+  /**
+   * Set a timeout for each test.
+   */
+  @Rule
+  public Timeout timeout = new Timeout(300000);
+  private MiniOzoneHAClusterImpl cluster;
+  private OzoneManager ozoneManager;
+  private ClientProtocol clientProtocol;
+  private static final String VOLUME_NAME = "vol-" + UUID.randomUUID();
+  private int fromLayoutVersion;
+  private OzoneManagerProtocol omClient;
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(TestOMBucketLayoutUpgrade.class);
+
+  /**
+   * Defines a "from" layout version to finalize from.
+   *
+   * @return
+   */
+  @Parameterized.Parameters
+  public static Collection<Object[]> data() {
+    return Arrays.asList(new Object[][]{
+        {INITIAL_VERSION}
+    });
+  }
+
+
+  public TestOMBucketLayoutUpgrade(OMLayoutFeature fromVersion) {
+    this.fromLayoutVersion = fromVersion.layoutVersion();
+  }
+
+  /**
+   * Create a MiniDFSCluster for testing.
+   */
+  @Before
+  public void setup() throws Exception {
+    org.junit.Assume.assumeTrue("Check if there is need to finalize.",
+        maxLayoutVersion() > fromLayoutVersion);
+
+    OzoneConfiguration conf = new OzoneConfiguration();
+    String omServiceId = UUID.randomUUID().toString();
+    cluster = (MiniOzoneHAClusterImpl) MiniOzoneCluster.newOMHABuilder(conf)
+        .setClusterId(UUID.randomUUID().toString())
+        .setScmId(UUID.randomUUID().toString())
+        .setOMServiceId(omServiceId)
+        .setNumOfOzoneManagers(3)
+        .setNumDatanodes(1)
```

Comment:
One minor thing I missed, we can use `.withoutDatanodes()` here to speed up execution since we are only doing metadata ops.

Reply:
I had tried that but it causes the cluster to not start up correctly for some reason.